File: sdk/typescript/tests/streaming.test.ts
// Streaming-helper tests for the TypeScript SDK. Like client.test.ts
// these run against an in-process http.Server that emits canned SSE
// frames; no real Go binary is spawned. Tests cover both the
// chat/messageStream raw-event surface and the protocol-agnostic
// iterStream + StreamChunk normalization.

import { afterAll, beforeAll, describe, expect as vexpect, it } from "vitest";
import { createServer, Server } from "node:http";
import { AddressInfo } from "node:net";

import {
  MockAgentClient,
  normalizeAnthropicStream,
  normalizeOpenAIStream,
  parseSSEFrame,
} from "../src/client.js";
import { StreamChunk } from "../src/types.js";

let server: Server;
let port: number;

const openaiChunks = [
  `data: ${JSON.stringify({ choices: [{ delta: { content: "a" } }] })}`,
  `data: ${JSON.stringify({ choices: [{ delta: { content: "b" } }] })}`,
  `data: ${JSON.stringify({ choices: [{ delta: {}, finish_reason: "stop" }] })}`,
  `data: [DONE]`,
];

const anthropicEvents = [
  { type: "message_start", message: { model: "claude-x" } },
  { type: "content_block_start", index: 0, content_block: { type: "text" } },
  {
    type: "content_block_delta",
    index: 0,
    delta: { type: "text_delta", text: "hi" },
  },
  { type: "content_block_stop", index: 0 },
  { type: "message_delta", delta: { stop_reason: "end_turn" } },
  { type: "message_stop" },
];

function writeSSE(res: import("node:http").ServerResponse, frames: string[]) {
  res.writeHead(200, {
    "Content-Type": "text/event-stream",
    "Cache-Control": "no-cache",
    Connection: "keep-alive",
  });
  for (const frame of frames) {
    res.write(frame + "\n\n");
  }
  res.end();
}

beforeAll(async () => {
  server = createServer((req, res) => {
    if (req.method === "POST" && req.url === "/v1/chat/completions") {
      writeSSE(res, openaiChunks);
      return;
    }
    if (req.method === "POST" && req.url === "/v1/messages") {
      const frames = anthropicEvents.map(
        (e) => `event: ${e.type}\ndata: ${JSON.stringify(e)}`,
      );
      writeSSE(res, frames);
      return;
    }
    res.writeHead(404);
    res.end();
  });
  await new Promise<void>((resolve) => server.listen(0, resolve));
  port = (server.address() as AddressInfo).port;
});

afterAll(async () => {
  await new Promise<void>((resolve) => server.close(() => resolve()));
});

// --- parseSSEFrame ---

describe("parseSSEFrame", () => {
  it("extracts event + data lines", () => {
    const got = parseSSEFrame("event: foo\ndata: hello");
    vexpect(got).toEqual({ event: "foo", data: "hello" });
  });

  it("joins multiple data lines with \\n", () => {
    const got = parseSSEFrame("data: line1\ndata: line2");
    vexpect(got).toEqual({ event: "", data: "line1\nline2" });
  });

  it("ignores comment lines", () => {
    const got = parseSSEFrame(": keepalive\ndata: x");
    vexpect(got).toEqual({ event: "", data: "x" });
  });

  it("returns null when no data line is present", () => {
    vexpect(parseSSEFrame("event: foo")).toBeNull();
  });
});

// --- raw streaming via http server ---

describe("MockAgentClient.chatStream", () => {
  it("yields parsed chunks and stops on [DONE]", async () => {
    const client = new MockAgentClient({ baseUrl: `http://localhost:${port}` });
    const events: any[] = [];
    for await (const chunk of client.chatStream([{ role: "user", content: "x" }])) {
      events.push(chunk);
    }
    vexpect(events.length).toBe(3);
    vexpect(events[0].choices[0].delta.content).toBe("a");
    vexpect(events[2].choices[0].finish_reason).toBe("stop");
  });
});

describe("MockAgentClient.messageStream", () => {
  it("yields anthropic events and stops on message_stop", async () => {
    const client = new MockAgentClient({ baseUrl: `http://localhost:${port}` });
    const events: any[] = [];
    for await (const event of client.messageStream([{ role: "user", content: "x" }])) {
      events.push(event);
    }
    const types = events.map((e) => e.type);
    vexpect(types).toEqual([
      "message_start",
      "content_block_start",
      "content_block_delta",
      "content_block_stop",
      "message_delta",
      "message_stop",
    ]);
  });
});

// --- iterStream end-to-end ---

describe("MockAgentClient.iterStream", () => {
  it("normalizes openai chunks to StreamChunks", async () => {
    const client = new MockAgentClient({ baseUrl: `http://localhost:${port}` });
    const chunks: StreamChunk[] = [];
    for await (const chunk of client.iterStream([{ role: "user", content: "x" }], {
      protocol: "openai",
    })) {
      chunks.push(chunk);
    }
    const text = chunks.map((c) => c.text).join("");
    vexpect(text).toBe("ab");
    vexpect(chunks[chunks.length - 1].finished).toBe(true);
    vexpect(chunks[chunks.length - 1].finishReason).toBe("stop");
  });

  it("normalizes anthropic events to StreamChunks", async () => {
    const client = new MockAgentClient({ baseUrl: `http://localhost:${port}` });
    const chunks: StreamChunk[] = [];
    for await (const chunk of client.iterStream([{ role: "user", content: "x" }], {
      protocol: "anthropic",
    })) {
      chunks.push(chunk);
    }
    const text = chunks.map((c) => c.text).join("");
    vexpect(text).toBe("hi");
    vexpect(chunks[chunks.length - 1].finished).toBe(true);
    vexpect(chunks[chunks.length - 1].finishReason).toBe("end_turn");
  });

  it("rejects unknown protocol", async () => {
    const client = new MockAgentClient({ baseUrl: `http://localhost:${port}` });
    await vexpect(async () => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const it = client.iterStream([], { protocol: "bogus" } as any);
      await it.next();
    }).rejects.toThrow(/unknown protocol/);
  });
});

// --- normalizer unit tests against constructed inputs ---

async function* fromArray<T>(items: T[]): AsyncGenerator<T> {
  for (const item of items) yield item;
}

describe("normalizeOpenAIStream", () => {
  it("filters padding chunks and emits tool call deltas", async () => {
    const raw = [
      { choices: [{ delta: {} }] }, // padding -> skipped
      { choices: [{ delta: { content: "x" } }] },
      {
        choices: [
          {
            delta: {
              tool_calls: [
                { index: 0, function: { name: "search", arguments: '{"q":' } },
              ],
            },
          },
        ],
      },
      { choices: [{ delta: {}, finish_reason: "tool_calls" }] },
    ];
    const out: StreamChunk[] = [];
    for await (const c of normalizeOpenAIStream(fromArray(raw))) out.push(c);
    vexpect(out.length).toBe(3);
    vexpect(out[0].text).toBe("x");
    vexpect(out[1].toolCallDelta).toEqual([0, "search", '{"q":']);
    vexpect(out[2].finished).toBe(true);
    vexpect(out[2].finishReason).toBe("tool_calls");
  });
});

describe("normalizeAnthropicStream", () => {
  it("emits text + finish for a text-only stream", async () => {
    const events = [
      { type: "message_start", message: {} },
      { type: "content_block_start", index: 0, content_block: { type: "text" } },
      {
        type: "content_block_delta",
        delta: { type: "text_delta", text: "hello " },
      },
      {
        type: "content_block_delta",
        delta: { type: "text_delta", text: "world" },
      },
      { type: "message_delta", delta: { stop_reason: "end_turn" } },
      { type: "message_stop" },
    ];
    const out: StreamChunk[] = [];
    for await (const c of normalizeAnthropicStream(fromArray(events))) out.push(c);
    const text = out.map((c) => c.text).join("");
    vexpect(text).toBe("hello world");
    vexpect(out[out.length - 1].finished).toBe(true);
    vexpect(out[out.length - 1].finishReason).toBe("end_turn");
  });

  it("accumulates input_json_delta fragments under the active tool", async () => {
    const events = [
      {
        type: "content_block_start",
        index: 0,
        content_block: { type: "tool_use", name: "get_weather" },
      },
      {
        type: "content_block_delta",
        index: 0,
        delta: { type: "input_json_delta", partial_json: '{"city":' },
      },
      {
        type: "content_block_delta",
        index: 0,
        delta: { type: "input_json_delta", partial_json: '"Tokyo"}' },
      },
      { type: "message_stop" },
    ];
    const out: StreamChunk[] = [];
    for await (const c of normalizeAnthropicStream(fromArray(events))) out.push(c);
    const fragments = out
      .filter((c) => c.toolCallDelta && c.toolCallDelta[2].length > 0)
      .map((c) => c.toolCallDelta![2])
      .join("");
    vexpect(fragments).toBe('{"city":"Tokyo"}');
    vexpect(out.find((c) => c.toolCallDelta && c.toolCallDelta[1] === "get_weather")).toBeTruthy();
    vexpect(out[out.length - 1].finished).toBe(true);
  });

  it("ignores events past message_stop", async () => {
    const events = [
      { type: "message_stop" },
      {
        type: "content_block_delta",
        delta: { type: "text_delta", text: "leak" },
      },
    ];
    const out: StreamChunk[] = [];
    for await (const c of normalizeAnthropicStream(fromArray(events))) out.push(c);
    vexpect(out.length).toBe(1);
    vexpect(out[0].finished).toBe(true);
  });
});

File: sdk/go/mockagents/streaming_test.go
package mockagents

import (
	"context"
	"fmt"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestParseSSEFrame(t *testing.T) {
	cases := []struct {
		name     string
		frame    string
		wantData string
		wantNil  bool
	}{
		{"basic", "event: delta\ndata: hello", "hello", false},
		{"no_leading_space", "data:raw", "raw", false},
		{"multi_data", "data: line1\ndata: line2", "line1\nline2", false},
		{"skip_comment", ":heartbeat\ndata: ok", "ok", false},
		{"no_data", "event: noop\n", "", true},
		{"crlf", "event: x\r\ndata: y\r\n", "y", false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := parseSSEFrame(tc.frame)
			if tc.wantNil {
				if got != nil {
					t.Fatalf("want nil frame, got %+v", got)
				}
				return
			}
			if got == nil {
				t.Fatal("unexpected nil frame")
			}
			if got.Data != tc.wantData {
				t.Errorf("data = %q, want %q", got.Data, tc.wantData)
			}
		})
	}
}

func TestNormalizeOpenAIStreamDropsPadding(t *testing.T) {
	var st normalizeState
	got := normalizeOpenAIStream(map[string]any{
		"choices": []any{map[string]any{"delta": map[string]any{}}},
	}, &st)
	if len(got) != 0 {
		t.Fatalf("padding chunk should drop, got %+v", got)
	}
}

func TestNormalizeOpenAIStreamTextAndToolAndFinish(t *testing.T) {
	var st normalizeState
	text := normalizeOpenAIStream(map[string]any{
		"choices": []any{map[string]any{
			"delta": map[string]any{"content": "hi"},
		}},
	}, &st)
	if len(text) != 1 || text[0].Text != "hi" || text[0].Finished {
		t.Errorf("text chunk wrong: %+v", text)
	}

	tool := normalizeOpenAIStream(map[string]any{
		"choices": []any{map[string]any{
			"delta": map[string]any{
				"tool_calls": []any{map[string]any{
					"index":    float64(2),
					"function": map[string]any{"name": "lookup", "arguments": `{"q":`},
				}},
			},
		}},
	}, &st)
	if len(tool) != 1 || tool[0].ToolCallDelta == nil {
		t.Fatalf("want tool delta, got %+v", tool)
	}
	if tool[0].ToolCallDelta.Index != 2 || tool[0].ToolCallDelta.Name != "lookup" || tool[0].ToolCallDelta.Fragment != `{"q":` {
		t.Errorf("tool delta = %+v", tool[0].ToolCallDelta)
	}

	done := normalizeOpenAIStream(map[string]any{
		"choices": []any{map[string]any{
			"delta":         map[string]any{},
			"finish_reason": "stop",
		}},
	}, &st)
	if len(done) != 1 || !done[0].Finished || done[0].FinishReason != "stop" {
		t.Errorf("finish chunk = %+v", done)
	}
}

func TestNormalizeAnthropicStreamTextAndToolAccumulation(t *testing.T) {
	var st normalizeState

	// Open a tool_use block.
	start := normalizeAnthropicStream(map[string]any{
		"type":  "content_block_start",
		"index": float64(0),
		"content_block": map[string]any{
			"type": "tool_use",
			"name": "search",
		},
	}, &st)
	if len(start) != 1 || start[0].ToolCallDelta == nil || start[0].ToolCallDelta.Name != "search" {
		t.Fatalf("start = %+v", start)
	}

	// input_json_delta should reuse the remembered index+name.
	frag := normalizeAnthropicStream(map[string]any{
		"type": "content_block_delta",
		"delta": map[string]any{
			"type":         "input_json_delta",
			"partial_json": `{"q":"cats"}`,
		},
	}, &st)
	if len(frag) != 1 || frag[0].ToolCallDelta == nil ||
		frag[0].ToolCallDelta.Index != 0 || frag[0].ToolCallDelta.Name != "search" ||
		frag[0].ToolCallDelta.Fragment != `{"q":"cats"}` {
		t.Errorf("frag = %+v", frag[0].ToolCallDelta)
	}

	// Text delta.
	text := normalizeAnthropicStream(map[string]any{
		"type": "content_block_delta",
		"delta": map[string]any{
			"type": "text_delta",
			"text": "hello",
		},
	}, &st)
	if len(text) != 1 || text[0].Text != "hello" {
		t.Errorf("text = %+v", text)
	}

	// message_delta carries stop reason.
	_ = normalizeAnthropicStream(map[string]any{
		"type":  "message_delta",
		"delta": map[string]any{"stop_reason": "tool_use"},
	}, &st)

	// message_stop terminal.
	stop := normalizeAnthropicStream(map[string]any{"type": "message_stop"}, &st)
	if len(stop) != 1 || !stop[0].Finished || stop[0].FinishReason != "tool_use" {
		t.Errorf("stop = %+v", stop)
	}
}

func TestNormalizeAnthropicStreamDefaultFinishReason(t *testing.T) {
	var st normalizeState
	stop := normalizeAnthropicStream(map[string]any{"type": "message_stop"}, &st)
	if len(stop) != 1 || stop[0].FinishReason != "end_turn" {
		t.Errorf("stop = %+v", stop)
	}
}

// --- End-to-end streaming against a fake SSE server ---

// sseWrite flushes the SSE frame to the response writer.
func sseWrite(w http.ResponseWriter, payload string) {
	_, _ = fmt.Fprintf(w, "data: %s\n\n", payload)
	if f, ok := w.(http.Flusher); ok {
		f.Flush()
	}
}

func newSSEServer(t *testing.T, path string, frames []string, terminator string) *httptest.Server {
	t.Helper()
	mux := http.NewServeMux()
	mux.HandleFunc(path, func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/event-stream")
		w.WriteHeader(200)
		for _, frame := range frames {
			sseWrite(w, frame)
		}
		if terminator != "" {
			sseWrite(w, terminator)
		}
	})
	return httptest.NewServer(mux)
}

func TestChatStreamEndToEnd(t *testing.T) {
	frames := []string{
		`{"choices":[{"delta":{"content":"hel"}}]}`,
		`{"choices":[{"delta":{"content":"lo"}}]}`,
		`{"choices":[{"delta":{},"finish_reason":"stop"}]}`,
	}
	srv := newSSEServer(t, "/v1/chat/completions", frames, "[DONE]")
	defer srv.Close()

	c := NewClient(ClientOptions{BaseURL: srv.URL})
	stream, err := c.ChatStream(context.Background(), []ChatMessage{{Role: "user", Content: "hi"}}, ChatOptions{})
	if err != nil {
		t.Fatalf("ChatStream: %v", err)
	}
	defer stream.Close()

	var events []map[string]any
	for stream.Next() {
		events = append(events, stream.Value())
	}
	if err := stream.Err(); err != nil {
		t.Fatalf("stream err: %v", err)
	}
	if len(events) != 3 {
		t.Fatalf("events = %d, want 3: %+v", len(events), events)
	}
}

func TestIterStreamOpenAIAssemblesText(t *testing.T) {
	frames := []string{
		`{"choices":[{"delta":{"content":"hel"}}]}`,
		`{"choices":[{"delta":{"content":"lo"}}]}`,
		`{"choices":[{"delta":{},"finish_reason":"stop"}]}`,
	}
	srv := newSSEServer(t, "/v1/chat/completions", frames, "[DONE]")
	defer srv.Close()

	c := NewClient(ClientOptions{BaseURL: srv.URL})
	stream, err := c.IterStream(context.Background(), []ChatMessage{{Role: "user", Content: "hi"}}, IterStreamOptions{})
	if err != nil {
		t.Fatalf("IterStream: %v", err)
	}
	defer stream.Close()

	var text strings.Builder
	var finished bool
	for stream.Next() {
		ch := stream.Value()
		text.WriteString(ch.Text)
		if ch.Finished {
			finished = true
		}
	}
	if err := stream.Err(); err != nil {
		t.Fatalf("err: %v", err)
	}
	if text.String() != "hello" {
		t.Errorf("text = %q", text.String())
	}
	if !finished {
		t.Error("never saw finished chunk")
	}
}

func TestMessageStreamStopsOnMessageStop(t *testing.T) {
	frames := []string{
		`{"type":"message_start"}`,
		`{"type":"content_block_delta","delta":{"type":"text_delta","text":"hi"}}`,
		`{"type":"message_delta","delta":{"stop_reason":"end_turn"}}`,
		`{"type":"message_stop"}`,
		// This event should never be surfaced — stream stops after message_stop.
		`{"type":"ghost"}`,
	}
	srv := newSSEServer(t, "/v1/messages", frames, "")
	defer srv.Close()

	c := NewClient(ClientOptions{BaseURL: srv.URL})
	stream, err := c.MessageStream(context.Background(), []ChatMessage{{Role: "user", Content: "hi"}}, MessageOptions{})
	if err != nil {
		t.Fatalf("MessageStream: %v", err)
	}
	defer stream.Close()

	var kinds []string
	for stream.Next() {
		ev := stream.Value()
		t, _ := ev["type"].(string)
		kinds = append(kinds, t)
	}
	if err := stream.Err(); err != nil {
		t.Fatalf("err: %v", err)
	}
	want := []string{"message_start", "content_block_delta", "message_delta", "message_stop"}
	if len(kinds) != len(want) {
		t.Fatalf("kinds = %v, want %v", kinds, want)
	}
	for i, k := range want {
		if kinds[i] != k {
			t.Errorf("kinds[%d] = %q, want %q", i, kinds[i], k)
		}
	}
}

func TestIterStreamAnthropicEndToEnd(t *testing.T) {
	frames := []string{
		`{"type":"content_block_delta","delta":{"type":"text_delta","text":"hel"}}`,
		`{"type":"content_block_delta","delta":{"type":"text_delta","text":"lo"}}`,
		`{"type":"message_delta","delta":{"stop_reason":"end_turn"}}`,
		`{"type":"message_stop"}`,
	}
	srv := newSSEServer(t, "/v1/messages", frames, "")
	defer srv.Close()

	c := NewClient(ClientOptions{BaseURL: srv.URL})
	stream, err := c.IterStream(context.Background(), []ChatMessage{{Role: "user", Content: "hi"}}, IterStreamOptions{Protocol: "anthropic"})
	if err != nil {
		t.Fatalf("IterStream: %v", err)
	}
	defer stream.Close()

	var text strings.Builder
	var finishReason string
	for stream.Next() {
		ch := stream.Value()
		text.WriteString(ch.Text)
		if ch.Finished {
			finishReason = ch.FinishReason
		}
	}
	if err := stream.Err(); err != nil {
		t.Fatalf("err: %v", err)
	}
	if text.String() != "hello" {
		t.Errorf("text = %q", text.String())
	}
	if finishReason != "end_turn" {
		t.Errorf("finish = %q", finishReason)
	}
}

func TestIterStreamUnknownProtocol(t *testing.T) {
	c := NewClient(ClientOptions{BaseURL: "http://nowhere.invalid"})
	_, err := c.IterStream(context.Background(), nil, IterStreamOptions{Protocol: "cohere"})
	if err == nil {
		t.Fatal("expected error for unknown protocol")
	}
	if !strings.Contains(err.Error(), "cohere") {
		t.Errorf("error = %v", err)
	}
}

func TestChatStreamHTTPErrorSurfacesAsHTTPError(t *testing.T) {
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadRequest)
		_, _ = w.Write([]byte(`{"error":"bad"}`))
	}))
	defer srv.Close()
	c := NewClient(ClientOptions{BaseURL: srv.URL})
	_, err := c.ChatStream(context.Background(), []ChatMessage{{Role: "user", Content: "x"}}, ChatOptions{})
	if err == nil {
		t.Fatal("expected error")
	}
	if herr, ok := err.(*HTTPError); !ok || herr.Status != 400 {
		t.Errorf("err = %v", err)
	}
}
